src/mjolnir/valhalla_add_predicted_traffic.cc

```diff
@@ -0,0 +1,381 @@
+          } break;
+          case 1:
+            if (ext == "constrained") {
+              traffic.constrained_flow_speed = std::stoi(t);
+              stat.constrained_count++;
+            } else if (ext == "freeflow") {
+              traffic.free_flow_speed = std::stoi(t);
+              stat.free_flow_count++;
+            }
+            break;
+        }
+        field_num++;
+      }
+
+      // assume that we are in the same tile.
+      // trying to save on looking up/finding a vector for a tile.
+      if (last_tile_id == tile_id) {
+        ts.emplace_back(traffic);
+      } else {
+        unique_data.lock.lock();
+        if (last_tile_id != kInvalidGraphId) {
+          if (it != unique_data.tile_speeds.end())
+            it->second = ts;
+          else
+            unique_data.tile_speeds.insert({last_tile_id, ts});
+        }
+        it = unique_data.tile_speeds.find(tile_id);
+        if (it != unique_data.tile_speeds.end()) {
+          ts = it->second;
+        } else
+          ts.clear();
+        unique_data.lock.unlock();
+        last_tile_id = tile_id;
+        ts.emplace_back(traffic);
+      }
+    }
+    if (last_tile_id != kInvalidGraphId) {
+      unique_data.lock.lock();
+      if (it != unique_data.tile_speeds.end())
+        it->second = ts;
+      else
+        unique_data.tile_speeds.insert({last_tile_id, ts});
+      unique_data.lock.unlock();
+    }
+  }
+  file.close();
+}
+
+/**
+ * Read both the constrained and freeflow speed CSV files
+ * We expect the files to be named as <quadtreeID>.constrained.csv and
+ * <quadtreeID>.freeflow.csv. (e.g., 1202021.constrained.csv and 1202021.freeflow.csv)
+ */
+void parse_traffic_tiles(const std::string& traffic_dir,
+                         std::mutex& lock,
+                         std::unordered_set<std::string>::const_iterator tile_start,
+                         std::unordered_set<std::string>::const_iterator tile_end,
+                         unique_data_t& unique_data,
+                         std::promise<stats>& result) {
+
+  // Iterate through the tiles and parse them
+  stats stat{};
+  for (; tile_start != tile_end; ++tile_start) {
+    ParseTrafficFile(traffic_dir, *tile_start, "constrained", unique_data, stat);
+    ParseTrafficFile(traffic_dir, *tile_start, "freeflow", unique_data, stat);
+  }
+  result.set_value(stat);
+}
+
+void update_valhalla_tiles(
+    const bpt::ptree& pt,
+    std::mutex& lock,
+    std::unordered_map<vb::GraphId, std::vector<TrafficSpeeds>>::const_iterator tile_start,
+    std::unordered_map<vb::GraphId, std::vector<TrafficSpeeds>>::const_iterator tile_end,
+    unique_data_t& unique_data,
+    std::promise<stats>& result) {
+
+  // Iterate through the tiles and parse them
+  stats stat{};
+  // Get the tile dir
+  std::string tile_dir = pt.get<std::string>("mjolnir.tile_dir");
+  // Iterate through the tiles in the queue and perform enhancements
+  for (; tile_start != tile_end; ++tile_start) {
+    lock.lock();
```

Comment:
If all of the inputs are broken into individual Valhalla tiles then you shouldn't need the lock.